R/pkg/R/SQLContext.R

```diff
@@ -90,19 +96,31 @@ createDataFrame <- function(sqlContext, data, schema = NULL, samplingRatio = 1.0
       if (is.null(schema)) {
         schema <- names(data)
       }
-      n <- nrow(data)
-      m <- ncol(data)
-      # get rid of factor type
-      dropFactor <- function(x) {
+
+      # types that should be wrapped before
+      # mapply to avoid flattening
+      requires_wrapping <- c("list")
+
+      # get rid of factor type and adjust for lists
+      cleanCols <- function(x) {
         if (is.factor(x)) {
           as.character(x)
+        } else if(class(x) %in% requires_wrapping) {
```

Comment:
The bug of SPARK-11283 lies in
```
lapply(1:m, function(j) { dropFactor(data[i,j]) })
```
in which data[i,j] returns a list of item instead item itself  when the item is in a column of list type.

Use mapply fixs it!